REGRESSION(287315@main): css-rhythm/parsing/block-step-computed crashing with ASSERT_NOT_IMPLEMENTED_YET

sammygill opened the pull request:
#### 4288ded87448ca4b0c1333f28590a946bf561f50
<pre>
REGRESSION(287315@main): css-rhythm/parsing/block-step-computed crashing with ASSERT_NOT_IMPLEMENTED_YET
<a href="https://bugs.webkit.org/show_bug.cgi?id=284021">https://bugs.webkit.org/show_bug.cgi?id=284021</a>
<a href="https://rdar.apple.com/140893010">rdar://140893010</a>

Reviewed by Tim Nguyen.

In 287315@main we did some refactoring and added some
ASSERT_NOT_IMPLEMENTED_YET()s on unsupported block-step-sizing content.
However, we failed to guard the call to performBlockStepSizing on
currently supported content so this results in block-step-computed.html
triggering the assert.

To fix this we can synchronize the logic used to guard the call
performBlockStepSizing and triggering the assert into a helper function.
We want to include a check to make sure block-step-insert is set to
margin-box since that is currently the only value we support for layout.

* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::BlockStepSizing::childHasSupportedStyle):
(WebCore::RenderBlockFlow::performBlockStepSizing const):
(WebCore::RenderBlockFlow::layoutBlockChild):

Canonical link: <a href="https://commits.webkit.org/287347@main">https://commits.webkit.org/287347@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6068ddcb4c49486e3b558ea145bb4aff782997cc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79219 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58247 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32591 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83841 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30388 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81352 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67339 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6517 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61971 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19874 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82286 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52025 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72183 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42276 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26328 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28774 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26757 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85236 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6517 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4513 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70215 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6680 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68054 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69464 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17326 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13516 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12337 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6474 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12250 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6398 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9865 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8190 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -941,9 +948,7 @@ void RenderBlockFlow::layoutInlineChildren(bool relayoutChildren, LayoutUnit& re
+    if (!BlockStepSizing::childHasSupportedStyle(child.style())) {
         ASSERT_NOT_IMPLEMENTED_YET();
         return;
     }
```
This should probably now be:
`ASSERT(BlockStepSizing::childHasSupportedStyle(child.style()))`.

since you can't really reach it otherwise.

sammygill replied:
Good point!

jyavenard commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -1022,7 +1024,8 @@ void RenderBlockFlow::layoutBlockChild(RenderBox& child, MarginInfo& marginInfo,
 
-    if (auto blockStepSizeForChild = child.style().blockStepSize())
+    auto& childStyle = child.style();
+    if (auto blockStepSizeForChild = childStyle.blockStepSize(); blockStepSizeForChild && BlockStepSizing::childHasSupportedStyle(child.style()))
```
why child.style() again if you took a reference earlier  which you also in that line ?

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -93,6 +93,13 @@ static_assert(sizeof(RenderBlockFlow::MarginInfo) == sizeof(SameSizeAsMarginInfo
 
 namespace BlockStepSizing {
 
+static bool childHasSupportedStyle(const RenderStyle& childStyle)
```
When I saw this function I was like this is confusingly vague and then I noticed the namespacing. I think in a long run it's better to have a dedicated helper class in a separate file as opposed to shovel all the things into RenderBlockFlow.cpp. It is already a confusing mix of loosely connected functionalities. TextBoxTrimmer is an example where the feature is limited to block containers yet it's not in RenderBlockFlow and does not require a namespace.

nt1m replied:
I did suggest moving this to a new file in https://github.com/WebKit/WebKit/pull/37351, but I think we should do that separately from fixing the crash. `TextBoxTrimmer` seems like a great pattern to follow.

sammygill replied:
Sure! While I somewhat had a preference initially, I don't feel that strongly about it!

sammygill replied:
(Will be moving this in a very soon to come follow up patch)

sammygill committed:
REGRESSION(287315@main): css-rhythm/parsing/block-step-computed crashing with ASSERT_NOT_IMPLEMENTED_YET
https://bugs.webkit.org/show_bug.cgi?id=284021
rdar://140893010

Reviewed by Tim Nguyen.

In 287315@main we did some refactoring and added some
ASSERT_NOT_IMPLEMENTED_YET()s on unsupported block-step-sizing content.
However, we failed to guard the call to performBlockStepSizing on
currently supported content so this results in block-step-computed.html
triggering the assert.

To fix this we can synchronize the logic used to guard the call
performBlockStepSizing and triggering the assert into a helper function.
We want to include a check to make sure block-step-insert is set to
margin-box since that is currently the only value we support for layout.

* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::BlockStepSizing::childHasSupportedStyle):
(WebCore::RenderBlockFlow::performBlockStepSizing const):
(WebCore::RenderBlockFlow::layoutBlockChild):

Canonical link: https://commits.webkit.org/287347@main
